Utilities/cmlibarchive/tar/bsdtar.1
.\" Copyright (c) 2003-2007 Tim Kientzle
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: src/usr.bin/tar/bsdtar.1,v 1.46 2008/12/06 07:37:55 kientzle Exp $
.\"
.Dd May 26, 2009
.Dt BSDTAR 1
.Os
.Sh NAME
.Nm tar
.Nd manipulate tape archives
.Sh SYNOPSIS
.Nm
.Op Ar bundled-flags Ao args Ac
.Op Ao Ar file Ac | Ao Ar pattern Ac ...
.Nm
.Brq Fl c
.Op Ar options
.Op Ar files | Ar directories
.Nm
.Brq Fl r | Fl u
.Fl f Ar archive-file
.Op Ar options
.Op Ar files | Ar directories
.Nm
.Brq Fl t | Fl x
.Op Ar options
.Op Ar patterns
.Sh DESCRIPTION
.Nm
creates and manipulates streaming archive files.
This implementation can extract from tar, pax, cpio, zip, jar, ar,
and ISO 9660 cdrom images and can create tar, pax, cpio, ar,
and shar archives.
.Pp
The first synopsis form shows a
.Dq bundled
option word.
This usage is provided for compatibility with historical implementations.
See COMPATIBILITY below for details.
.Pp
The other synopsis forms show the preferred usage.
The first option to
.Nm
is a mode indicator from the following list:
.Bl -tag -compact -width indent
.It Fl c
Create a new archive containing the specified items.
.It Fl r
Like
.Fl c ,
but new entries are appended to the archive.
Note that this only works on uncompressed archives stored in regular files.
The
.Fl f
option is required.
.It Fl t
List archive contents to stdout.
.It Fl u
Like
.Fl r ,
but new entries are added only if they have a modification date
newer than the corresponding entry in the archive.
Note that this only works on uncompressed archives stored in regular files.
The
.Fl f
option is required.
.It Fl x
Extract to disk from the archive.
If a file with the same name appears more than once in the archive,
each copy will be extracted, with later copies overwriting (replacing)
earlier copies.
.El
.Pp
In
.Fl c ,
.Fl r ,
or
.Fl u
mode, each specified file or directory is added to the
archive in the order specified on the command line.
By default, the contents of each directory are also archived.
.Pp
In extract or list mode, the entire command line
is read and parsed before the archive is opened.
The pathnames or patterns on the command line indicate
which items in the archive should be processed.
Patterns are shell-style globbing patterns as
documented in
.Xr tcsh 1 .
.Sh OPTIONS
Unless specifically stated otherwise, options are applicable in
all operating modes.
.Bl -tag -width indent
.It Cm @ Ns Pa archive
(c and r mode only)
The specified archive is opened and the entries
in it will be appended to the current archive.
As a simple example,
.Dl Nm Fl c Fl f Pa - Pa newfile Cm @ Ns Pa original.tar
writes a new archive to standard output containing a file
.Pa newfile
and all of the entries from
.Pa original.tar .
In contrast,
.Dl Nm Fl c Fl f Pa - Pa newfile Pa original.tar
creates a new archive with only two entries.
Similarly,
.Dl Nm Fl czf Pa - Fl -format Cm pax Cm @ Ns Pa -
reads an archive from standard input (whose format will be determined
automatically) and converts it into a gzip-compressed
pax-format archive on stdout.
In this way,
.Nm
can be used to convert archives from one format to another.
.It Fl b Ar blocksize
Specify the block size, in 512-byte records, for tape drive I/O.
As a rule, this argument is only needed when reading from or writing
to tape drives, and usually not even then as the default block size of
20 records (10240 bytes) is very common.
.It Fl C Ar directory
In c and r mode, this changes the directory before adding
the following files.
In x mode, change directories after opening the archive
but before extracting entries from the archive.
.It Fl -check-links
(c and r modes only)
Issue a warning message unless all links to each file are archived.
.It Fl -chroot
(x mode only)
.Fn chroot
to the current directory after processing any
.Fl C
options and before extracting any files.
.It Fl -exclude Ar pattern
Do not process files or directories that match the
specified pattern.
Note that exclusions take precedence over patterns or filenames
specified on the command line.
.It Fl -format Ar format
(c, r, u mode only)
Use the specified format for the created archive.
Supported formats include
.Dq cpio ,
.Dq pax ,
.Dq shar ,
and
.Dq ustar .
Other formats may also be supported; see
.Xr libarchive-formats 5
for more information about currently-supported formats.
In r and u modes, when extending an existing archive, the format specified
here must be compatible with the format of the existing archive on disk.
.It Fl f Ar file
Read the archive from or write the archive to the specified file.
The filename can be
.Pa -
for standard input or standard output.
If not specified, the default tape device will be used.
(On
.Fx ,
the default tape device is
.Pa /dev/sa0 . )
.It Fl H
(c and r mode only)
Symbolic links named on the command line will be followed; the
target of the link will be archived, not the link itself.
.It Fl h
(c and r mode only)
Synonym for
.Fl L .
.It Fl I
Synonym for
.Fl T .
.It Fl -include Ar pattern
Process only files or directories that match the specified pattern.
Note that exclusions specified with
.Fl -exclude
take precedence over inclusions.
If no inclusions are explicitly specified, all entries are processed by
default.
The
.Fl -include
option is especially useful when filtering archives.
For example, the command
.Dl Nm Fl c Fl f Pa new.tar Fl -include='*foo*' Cm @ Ns Pa old.tgz
creates a new archive
.Pa new.tar
containing only the entries from
.Pa old.tgz
containing the string
.Sq foo .
.It Fl j
(c mode only)
Compress the resulting archive with
.Xr bzip2 1 .
In extract or list modes, this option is ignored.
Note that, unlike other
.Nm tar
implementations, this implementation recognizes bzip2 compression
automatically when reading archives.
.It Fl k
(x mode only)
Do not overwrite existing files.
In particular, if a file appears more than once in an archive,
later copies will not overwrite earlier copies.
.It Fl -keep-newer-files
(x mode only)
Do not overwrite existing files that are newer than the
versions appearing in the archive being extracted.
.It Fl L
(c and r mode only)
All symbolic links will be followed.
Normally, symbolic links are archived as such.
With this option, the target of the link will be archived instead.
.It Fl l
This is a synonym for the
.Fl -check-links
option.
.It Fl m
(x mode only)
Do not extract modification time.
By default, the modification time is set to the time stored in the archive.
.It Fl n
(c, r, u modes only)
Do not recursively archive the contents of directories.
.It Fl -newer Ar date
(c, r, u modes only)
Only include files and directories newer than the specified date.
This compares ctime entries.
.It Fl -newer-mtime Ar date
(c, r, u modes only)
Like
.Fl -newer ,
except it compares mtime entries instead of ctime entries.
.It Fl -newer-than Pa file
(c, r, u modes only)
Only include files and directories newer than the specified file.
This compares ctime entries.
.It Fl -newer-mtime-than Pa file
(c, r, u modes only)
Like
.Fl -newer-than ,
except it compares mtime entries instead of ctime entries.
.It Fl -nodump
(c and r modes only)
Honor the nodump file flag by skipping this file.
.It Fl -null
(use with
.Fl I ,
.Fl T ,
or
.Fl X )
Filenames or patterns are separated by null characters,
not by newlines.
This is often used to read filenames output by the
.Fl print0
option to
.Xr find 1 .
.It Fl -numeric-owner
(x mode only)
Ignore symbolic user and group names when restoring archives to disk,
only numeric uid and gid values will be obeyed.
.It Fl O
(x, t modes only)
In extract (-x) mode, files will be written to standard out rather than
being extracted to disk.
In list (-t) mode, the file listing will be written to stderr rather than
the usual stdout.
.It Fl o
(x mode)
Use the user and group of the user running the program rather
than those specified in the archive.
Note that this has no significance unless
.Fl p
is specified, and the program is being run by the root user.
In this case, the file modes and flags from
the archive will be restored, but ACLs or owner information in
the archive will be discarded.
.It Fl o
(c, r, u mode)
A synonym for
.Fl -format Ar ustar
.It Fl -one-file-system
(c, r, and u modes)
Do not cross mount points.
.It Fl -options Ar options
Select optional behaviors for particular modules.

Utilities/cmlibarchive/tar/bsdtar.c
/*-
 * Copyright (c) 2003-2008 Tim Kientzle
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR(S) ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR(S) BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "bsdtar_platform.h"
__FBSDID("$FreeBSD: src/usr.bin/tar/bsdtar.c,v 1.93 2008/11/08 04:43:24 kientzle Exp $");

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#ifdef HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifdef HAVE_LANGINFO_H
#include <langinfo.h>
#endif
#ifdef HAVE_LOCALE_H
#include <locale.h>
#endif
#ifdef HAVE_PATHS_H
#include <paths.h>
#endif
#ifdef HAVE_SIGNAL_H
#include <signal.h>
#endif
#include <stdio.h>
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_STRING_H
#include <string.h>
#endif
#ifdef HAVE_TIME_H
#include <time.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#if HAVE_ZLIB_H
#include <cm_zlib.h>
#endif

#include "bsdtar.h"
#include "err.h"

/*
 * Per POSIX.1-1988, tar defaults to reading/writing archives to/from
 * the default tape device for the system.  Pick something reasonable here.
 */
#ifdef __linux
#define _PATH_DEFTAPE "/dev/st0"
#endif
#if defined(_WIN32) && !defined(__CYGWIN__)
#define _PATH_DEFTAPE "\\\\.\\tape0"
#endif

#ifndef _PATH_DEFTAPE
#define _PATH_DEFTAPE "/dev/tape"
#endif

static struct bsdtar *_bsdtar;

#if defined(SIGINFO) || defined(SIGUSR1)
static volatile int siginfo_occurred;

static void
siginfo_handler(int sig)
{
    (void)sig; /* UNUSED */
    siginfo_occurred = 1;
}

int
need_report(void)
{
    int r = siginfo_occurred;
    siginfo_occurred = 0;
    return (r);
}
#else
int
need_report(void)
{
    return (0);
}
#endif

/* External function to parse a date/time string */
time_t get_date(time_t, const char *);

static void      long_help(void);
static void      only_mode(struct bsdtar *, const char *opt,
                 const char *valid);
static void      set_mode(struct bsdtar *, char opt);
static void      version(void);

/* A basic set of security flags to request from libarchive. */
#define SECURITY                    \
    (ARCHIVE_EXTRACT_SECURE_SYMLINKS        \
     | ARCHIVE_EXTRACT_SECURE_NODOTDOT)

int
main(int argc, char **argv)
{
    struct bsdtar       *bsdtar, bsdtar_storage;
    int          opt, t;
    char             option_o;
    char             possible_help_request;
    char             buff[16];
    time_t           now;

    /*
     * Use a pointer for consistency, but stack-allocated storage
     * for ease of cleanup.
     */
    _bsdtar = bsdtar = &bsdtar_storage;
    memset(bsdtar, 0, sizeof(*bsdtar));
    bsdtar->fd = -1; /* Mark as "unused" */
    option_o = 0;

#if defined(SIGINFO) || defined(SIGUSR1)
    { /* Catch SIGINFO and SIGUSR1, if they exist. */
        struct sigaction sa;
        sa.sa_handler = siginfo_handler;
        sigemptyset(&sa.sa_mask);
        sa.sa_flags = 0;
#ifdef SIGINFO
        if (sigaction(SIGINFO, &sa, NULL))
            lafe_errc(1, errno, "sigaction(SIGINFO) failed");
#endif
#ifdef SIGUSR1
        /* ... and treat SIGUSR1 the same way as SIGINFO. */
        if (sigaction(SIGUSR1, &sa, NULL))
            lafe_errc(1, errno, "sigaction(SIGUSR1) failed");
#endif
    }
#endif


    /* Need lafe_progname before calling lafe_warnc. */
    if (*argv == NULL)
        lafe_progname = "bsdtar";
    else {
#if defined(_WIN32) && !defined(__CYGWIN__)
        lafe_progname = strrchr(*argv, '\\');
#else
        lafe_progname = strrchr(*argv, '/');
#endif
        if (lafe_progname != NULL)
            lafe_progname++;
        else
            lafe_progname = *argv;
    }

    time(&now);

#if HAVE_SETLOCALE
    if (setlocale(LC_ALL, "") == NULL)
        lafe_warnc(0, "Failed to set default locale");
#endif
#if defined(HAVE_NL_LANGINFO) && defined(HAVE_D_MD_ORDER)
    bsdtar->day_first = (*nl_langinfo(D_MD_ORDER) == 'd');
#endif
    possible_help_request = 0;

    /* Look up uid of current user for future reference */
    bsdtar->user_uid = geteuid();

    /* Default: open tape drive. */
    bsdtar->filename = getenv("TAPE");
    if (bsdtar->filename == NULL)
        bsdtar->filename = _PATH_DEFTAPE;

    /* Default: preserve mod time on extract */
    bsdtar->extract_flags = ARCHIVE_EXTRACT_TIME;

    /* Default: Perform basic security checks. */
    bsdtar->extract_flags |= SECURITY;

#ifndef _WIN32
    /* On POSIX systems, assume --same-owner and -p when run by
     * the root user.  This doesn't make any sense on Windows. */
    if (bsdtar->user_uid == 0) {
        /* --same-owner */
        bsdtar->extract_flags |= ARCHIVE_EXTRACT_OWNER;
        /* -p */
        bsdtar->extract_flags |= ARCHIVE_EXTRACT_PERM;
        bsdtar->extract_flags |= ARCHIVE_EXTRACT_ACL;
        bsdtar->extract_flags |= ARCHIVE_EXTRACT_XATTR;
        bsdtar->extract_flags |= ARCHIVE_EXTRACT_FFLAGS;
    }
#endif

    bsdtar->argv = argv;
    bsdtar->argc = argc;

    /*
     * Comments following each option indicate where that option
     * originated:  SUSv2, POSIX, GNU tar, star, etc.  If there's
     * no such comment, then I don't know of anyone else who
     * implements that option.
     */
    while ((opt = bsdtar_getopt(bsdtar)) != -1) {
        switch (opt) {
        case 'B': /* GNU tar */
            /* libarchive doesn't need this; just ignore it. */
            break;
        case 'b': /* SUSv2 */
            t = atoi(bsdtar->optarg);
            if (t <= 0 || t > 1024)
                lafe_errc(1, 0,
                    "Argument to -b is out of range (1..1024)");
            bsdtar->bytes_per_block = 512 * t;
            break;
        case 'C': /* GNU tar */
            set_chdir(bsdtar, bsdtar->optarg);
            break;
        case 'c': /* SUSv2 */
            set_mode(bsdtar, opt);
            break;
        case OPTION_CHECK_LINKS: /* GNU tar */
            bsdtar->option_warn_links = 1;
            break;
        case OPTION_CHROOT: /* NetBSD */
            bsdtar->option_chroot = 1;
            break;
        case OPTION_EXCLUDE: /* GNU tar */
            if (lafe_exclude(&bsdtar->matching, bsdtar->optarg))
                lafe_errc(1, 0,
                    "Couldn't exclude %s\n", bsdtar->optarg);
            break;
        case OPTION_FORMAT: /* GNU tar, others */
            bsdtar->create_format = bsdtar->optarg;
            break;
        case OPTION_OPTIONS:
            bsdtar->option_options = bsdtar->optarg;
            break;
        case 'f': /* SUSv2 */
            bsdtar->filename = bsdtar->optarg;
            if (strcmp(bsdtar->filename, "-") == 0)
                bsdtar->filename = NULL;
            break;
        case 'H': /* BSD convention */
            bsdtar->symlink_mode = 'H';
            break;
        case 'h': /* Linux Standards Base, gtar; synonym for -L */
            bsdtar->symlink_mode = 'L';
            /* Hack: -h by itself is the "help" command. */
            possible_help_request = 1;
            break;
        case OPTION_HELP: /* GNU tar, others */
            long_help();
            exit(0);
            break;
        case 'I': /* GNU tar */
            /*
             * TODO: Allow 'names' to come from an archive,
             * not just a text file.  Design a good UI for
             * allowing names and mode/owner to be read
             * from an archive, with contents coming from
             * disk.  This can be used to "refresh" an
             * archive or to design archives with special
             * permissions without having to create those
             * permissions on disk.
             */
            bsdtar->names_from_file = bsdtar->optarg;
            break;
        case OPTION_INCLUDE:
            /*
             * Noone else has the @archive extension, so
             * noone else needs this to filter entries
             * when transforming archives.
             */
            if (lafe_include(&bsdtar->matching, bsdtar->optarg))
                lafe_errc(1, 0,
                    "Failed to add %s to inclusion list",
                    bsdtar->optarg);
            break;
        case 'j': /* GNU tar */
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case 'J': /* GNU tar 1.21 and later */
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case 'k': /* GNU tar */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_NO_OVERWRITE;
            break;
        case OPTION_KEEP_NEWER_FILES: /* GNU tar */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_NO_OVERWRITE_NEWER;
            break;
        case 'L': /* BSD convention */
            bsdtar->symlink_mode = 'L';
            break;
            case 'l': /* SUSv2 and GNU tar beginning with 1.16 */
            /* GNU tar 1.13  used -l for --one-file-system */
            bsdtar->option_warn_links = 1;
            break;
        case OPTION_LZMA:
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case 'm': /* SUSv2 */
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_TIME;
            break;
        case 'n': /* GNU tar */
            bsdtar->option_no_subdirs = 1;
            break;
            /*
         * Selecting files by time:
         *    --newer-?time='date' Only files newer than 'date'
         *    --newer-?time-than='file' Only files newer than time
         *         on specified file (useful for incremental backups)
         * TODO: Add corresponding "older" options to reverse these.
         */
        case OPTION_NEWER_CTIME: /* GNU tar */
            bsdtar->newer_ctime_sec = get_date(now, bsdtar->optarg);
            break;
        case OPTION_NEWER_CTIME_THAN:
            {
                struct stat st;
                if (stat(bsdtar->optarg, &st) != 0)
                    lafe_errc(1, 0,
                        "Can't open file %s", bsdtar->optarg);
                bsdtar->newer_ctime_sec = st.st_ctime;
                bsdtar->newer_ctime_nsec =
                    ARCHIVE_STAT_CTIME_NANOS(&st);
            }
            break;
        case OPTION_NEWER_MTIME: /* GNU tar */
            bsdtar->newer_mtime_sec = get_date(now, bsdtar->optarg);
            break;
        case OPTION_NEWER_MTIME_THAN:
            {
                struct stat st;
                if (stat(bsdtar->optarg, &st) != 0)
                    lafe_errc(1, 0,
                        "Can't open file %s", bsdtar->optarg);
                bsdtar->newer_mtime_sec = st.st_mtime;
                bsdtar->newer_mtime_nsec =
                    ARCHIVE_STAT_MTIME_NANOS(&st);
            }
            break;
        case OPTION_NODUMP: /* star */
            bsdtar->option_honor_nodump = 1;
            break;
        case OPTION_NO_SAME_OWNER: /* GNU tar */
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_OWNER;
            break;
        case OPTION_NO_SAME_PERMISSIONS: /* GNU tar */
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_PERM;
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_ACL;
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_XATTR;
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_FFLAGS;
            break;
        case OPTION_NULL: /* GNU tar */
            bsdtar->option_null++;
            break;
        case OPTION_NUMERIC_OWNER: /* GNU tar */
            bsdtar->option_numeric_owner++;
            break;
        case 'O': /* GNU tar */
            bsdtar->option_stdout = 1;
            break;
        case 'o': /* SUSv2 and GNU conflict here, but not fatally */
            option_o = 1; /* Record it and resolve it later. */
            break;
        case OPTION_ONE_FILE_SYSTEM: /* GNU tar */
            bsdtar->option_dont_traverse_mounts = 1;
            break;
#if 0
        /*
         * The common BSD -P option is not necessary, since
         * our default is to archive symlinks, not follow
         * them.  This is convenient, as -P conflicts with GNU
         * tar anyway.
         */
        case 'P': /* BSD convention */
            /* Default behavior, no option necessary. */
            break;
#endif
        case 'P': /* GNU tar */
            bsdtar->extract_flags &= ~SECURITY;
            bsdtar->option_absolute_paths = 1;
            break;
        case 'p': /* GNU tar, star */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_PERM;
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_ACL;
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_XATTR;
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_FFLAGS;
            break;
        case OPTION_POSIX: /* GNU tar */
            bsdtar->create_format = "pax";
            break;
        case 'q': /* FreeBSD GNU tar --fast-read, NetBSD -q */
            bsdtar->option_fast_read = 1;
            break;
        case 'r': /* SUSv2 */
            set_mode(bsdtar, opt);
            break;
        case 'S': /* NetBSD pax-as-tar */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_SPARSE;
            break;
        case 's': /* NetBSD pax-as-tar */
#if HAVE_REGEX_H
            add_substitution(bsdtar, bsdtar->optarg);
#else
            lafe_warnc(0,
                "-s is not supported by this version of bsdtar");
            usage();
#endif
            break;
        case OPTION_SAME_OWNER: /* GNU tar */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_OWNER;
            break;
        case OPTION_STRIP_COMPONENTS: /* GNU tar 1.15 */
            bsdtar->strip_components = atoi(bsdtar->optarg);
            break;
        case 'T': /* GNU tar */
            bsdtar->names_from_file = bsdtar->optarg;
            break;
        case 't': /* SUSv2 */
            set_mode(bsdtar, opt);
            bsdtar->verbose++;
            break;
        case OPTION_TOTALS: /* GNU tar */
            bsdtar->option_totals++;
            break;
        case 'U': /* GNU tar */
            bsdtar->extract_flags |= ARCHIVE_EXTRACT_UNLINK;
            bsdtar->option_unlink_first = 1;
            break;
        case 'u': /* SUSv2 */
            set_mode(bsdtar, opt);
            break;
        case 'v': /* SUSv2 */
            bsdtar->verbose++;
            break;
        case OPTION_VERSION: /* GNU convention */
            version();
            break;
#if 0
        /*
         * The -W longopt feature is handled inside of
         * bsdtar_getopt(), so -W is not available here.
         */
        case 'W': /* Obscure GNU convention. */
            break;
#endif
        case 'w': /* SUSv2 */
            bsdtar->option_interactive = 1;
            break;
        case 'X': /* GNU tar */
            if (lafe_exclude_from_file(&bsdtar->matching, bsdtar->optarg))
                lafe_errc(1, 0,
                    "failed to process exclusions from file %s",
                    bsdtar->optarg);
            break;
        case 'x': /* SUSv2 */
            set_mode(bsdtar, opt);
            break;
        case 'y': /* FreeBSD version of GNU tar */
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case 'Z': /* GNU tar */
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case 'z': /* GNU tar, star, many others */
            if (bsdtar->create_compression != '\0')
                lafe_errc(1, 0,
                    "Can't specify both -%c and -%c", opt,
                    bsdtar->create_compression);
            bsdtar->create_compression = opt;
            break;
        case OPTION_USE_COMPRESS_PROGRAM:
            bsdtar->compress_program = bsdtar->optarg;
            break;
        default:
            usage();
        }
    }

    /*
     * Sanity-check options.
     */

    /* If no "real" mode was specified, treat -h as --help. */
    if ((bsdtar->mode == '\0') && possible_help_request) {
        long_help();
        exit(0);
    }

    /* Otherwise, a mode is required. */
    if (bsdtar->mode == '\0')
        lafe_errc(1, 0,
            "Must specify one of -c, -r, -t, -u, -x");

    /* Check boolean options only permitted in certain modes. */
    if (bsdtar->option_dont_traverse_mounts)
        only_mode(bsdtar, "--one-file-system", "cru");
    if (bsdtar->option_fast_read)
        only_mode(bsdtar, "--fast-read", "xt");
    if (bsdtar->option_honor_nodump)
        only_mode(bsdtar, "--nodump", "cru");
    if (option_o > 0) {
        switch (bsdtar->mode) {
        case 'c':
            /*
             * In GNU tar, -o means "old format."  The
             * "ustar" format is the closest thing
             * supported by libarchive.
             */
            bsdtar->create_format = "ustar";
            /* TODO: bsdtar->create_format = "v7"; */
            break;
        case 'x':
            /* POSIX-compatible behavior. */
            bsdtar->option_no_owner = 1;
            bsdtar->extract_flags &= ~ARCHIVE_EXTRACT_OWNER;
            break;
        default:
            only_mode(bsdtar, "-o", "xc");
            break;
        }
    }
    if (bsdtar->option_no_subdirs)
        only_mode(bsdtar, "-n", "cru");
    if (bsdtar->option_stdout)
        only_mode(bsdtar, "-O", "xt");
    if (bsdtar->option_unlink_first)
        only_mode(bsdtar, "-U", "x");
    if (bsdtar->option_warn_links)
        only_mode(bsdtar, "--check-links", "cr");

    /* Check other parameters only permitted in certain modes. */
    if (bsdtar->create_compression != '\0') {
        strcpy(buff, "-?");
        buff[1] = bsdtar->create_compression;
        only_mode(bsdtar, buff, "cxt");
    }
    if (bsdtar->create_format != NULL)
        only_mode(bsdtar, "--format", "cru");
    if (bsdtar->symlink_mode != '\0') {
        strcpy(buff, "-?");
        buff[1] = bsdtar->symlink_mode;
        only_mode(bsdtar, buff, "cru");
    }
    if (bsdtar->strip_components != 0)
        only_mode(bsdtar, "--strip-components", "xt");

    switch(bsdtar->mode) {
    case 'c':
        tar_mode_c(bsdtar);
        break;
    case 'r':
        tar_mode_r(bsdtar);
        break;
    case 't':
        tar_mode_t(bsdtar);
        break;
    case 'u':
        tar_mode_u(bsdtar);
        break;
    case 'x':
        tar_mode_x(bsdtar);
        break;
    }

    lafe_cleanup_exclusions(&bsdtar->matching);
#if HAVE_REGEX_H
    cleanup_substitution(bsdtar);
#endif

    if (bsdtar->return_value != 0)
        lafe_warnc(0,
            "Error exit delayed from previous errors.");
    return (bsdtar->return_value);
}

static void
set_mode(struct bsdtar *bsdtar, char opt)
{
    if (bsdtar->mode != '\0' && bsdtar->mode != opt)
        lafe_errc(1, 0,
            "Can't specify both -%c and -%c", opt, bsdtar->mode);
    bsdtar->mode = opt;
}

/*
 * Verify that the mode is correct.
 */
static void
only_mode(struct bsdtar *bsdtar, const char *opt, const char *valid_modes)
{
    if (strchr(valid_modes, bsdtar->mode) == NULL)
        lafe_errc(1, 0,
            "Option %s is not permitted in mode -%c",
            opt, bsdtar->mode);
}


void
usage(void)
{
    const char  *p;

    p = lafe_progname;

    fprintf(stderr, "Usage:\n");
    fprintf(stderr, "  List:    %s -tf <archive-filename>\n", p);
    fprintf(stderr, "  Extract: %s -xf <archive-filename>\n", p);
    fprintf(stderr, "  Create:  %s -cf <archive-filename> [filenames...]\n", p);
    fprintf(stderr, "  Help:    %s --help\n", p);
    exit(1);
}

static void
version(void)
{
    printf("bsdtar %s - %s\n",
        BSDTAR_VERSION_STRING,
        archive_version());
    exit(0);
}

static const char *long_help_msg =
    "First option must be a mode specifier:\n"
    "  -c Create  -r Add/Replace  -t List  -u Update  -x Extract\n"
    "Common Options:\n"
    "  -b #  Use # 512-byte records per I/O block\n"
    "  -f <filename>  Location of archive (default " _PATH_DEFTAPE ")\n"
    "  -v    Verbose\n"
    "  -w    Interactive\n"
    "Create: %p -c [options] [<file> | <dir> | @<archive> | -C <dir> ]\n"
    "  <file>, <dir>  add these items to archive\n"
    "  -z, -j, -J, --lzma  Compress archive with gzip/bzip2/xz/lzma\n"
    "  --format {ustar|pax|cpio|shar}  Select archive format\n"
    "  --exclude <pattern>  Skip files that match pattern\n"
    "  -C <dir>  Change to <dir> before processing remaining files\n"
    "  @<archive>  Add entries from <archive> to output\n"
    "List: %p -t [options] [<patterns>]\n"
    "  <patterns>  If specified, list only entries that match\n"
    "Extract: %p -x [options] [<patterns>]\n"
    "  <patterns>  If specified, extract only entries that match\n"
    "  -k    Keep (don't overwrite) existing files\n"
    "  -m    Don't restore modification times\n"
    "  -O    Write entries to stdout, don't restore to disk\n"
    "  -p    Restore permissions (including ACLs, owner, file flags)\n";


/*
 * Note that the word 'bsdtar' will always appear in the first line
 * of output.
 *
 * In particular, /bin/sh scripts that need to test for the presence
 * of bsdtar can use the following template:
 *
 * if (tar --help 2>&1 | grep bsdtar >/dev/null 2>&1 ) then \
 *          echo bsdtar; else echo not bsdtar; fi
 */
static void
long_help(void)
{
    const char  *prog;
    const char  *p;

    prog = lafe_progname;

    fflush(stderr);

    p = (strcmp(prog,"bsdtar") != 0) ? "(bsdtar)" : "";
    printf("%s%s: manipulate archive files\n", prog, p);

    for (p = long_help_msg; *p != '\0'; p++) {
        if (*p == '%') {
            if (p[1] == 'p') {
                fputs(prog, stdout);
                p++;
            } else
                putchar('%');
        } else
            putchar(*p);
    }
    version();
}
